pkg/ipfs/repository.go

```diff
@@ -0,0 +1,142 @@
+// VulcanizeDB
+// Copyright © 2019 Vulcanize
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package ipfs
+
+import (
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq"
+
+	"github.com/vulcanize/vulcanizedb/pkg/datastore/postgres"
+)
+
+// CIDRepository is an interface for indexing CIDPayloads
+type CIDRepository interface {
+	Index(cidPayload *CIDPayload) error
+}
+
+// Repository is the underlying struct for the CIDRepository interface
+type Repository struct {
+	db *postgres.DB
+}
+
+// NewCIDRepository creates a new pointer to a Repository which satisfies the CIDRepository interface
+func NewCIDRepository(db *postgres.DB) *Repository {
+	return &Repository{
+		db: db,
+	}
+}
+
+// Index indexes a cidPayload in Postgres
+func (repo *Repository) Index(cidPayload *CIDPayload) error {
+	tx, err := repo.db.Beginx()
+	if err != nil {
+		return err
+	}
+	headerID, err := repo.indexHeaderCID(tx, cidPayload.HeaderCID, cidPayload.BlockNumber, cidPayload.BlockHash.Hex())
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	for uncleHash, cid := range cidPayload.UncleCIDS {
+		err = repo.indexUncleCID(tx, cid, cidPayload.BlockNumber, uncleHash.Hex())
+		if err != nil {
+			tx.Rollback()
+			return err
+		}
+	}
+	err = repo.indexTransactionAndReceiptCIDs(tx, cidPayload, headerID)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	err = repo.indexStateAndStorageCIDs(tx, cidPayload, headerID)
+	if err != nil {
+		tx.Rollback()
+		return err
+	}
+	return tx.Commit()
+}
+
+func (repo *Repository) indexHeaderCID(tx *sqlx.Tx, cid, blockNumber, hash string) (int64, error) {
+	var headerID int64
+	err := tx.QueryRowx(`INSERT INTO public.header_cids (block_number, block_hash, cid, final) VALUES ($1, $2, $3, $4)
+								ON CONFLICT (block_number, block_hash) DO UPDATE SET (cid, final) = ($3, $4)
+								RETURNING id`,
+		blockNumber, hash, cid, true).Scan(&headerID)
```

Comment:
so is `final` on the headers table a way of distinguishing between headers and uncles? Does it get swapped if a header is reorg'd?

Reply:
If `final` is true then the header is canonical, if it is false then it is an uncle or an otherwise non-canonical header. If a reorg occurs then the new header will be processed and added to the database, if the reorged header has the same block number and hash then it will overwrite the previous one, but if it has a different hash (which I think reorged headers do by definition?) then the original non-final header will remain in the db as non-final.

Reply:
At least this is my understanding!

Reply:
A canonical hash is one that's older than 2048 blocks and found in the Canonical Hash Tree. Technically anything before that can be re-orged, including what the geth client calls `final`.

`final` hashes are NOT `canonical`!

In vulcanizeDB we shouldn't be talking about uncles, only Canonical hashes and the weights of other hashes. Since we will be getting hashes from multiple clients which will disagree if a hash is the uncle or not.

Reply:
Understood, this will need to be reworked.